src/nrnpython/nrnpy_hoc.cpp

```diff
@@ -219,37 +230,48 @@ static PyObject* hocobj_new(PyTypeObject* subtype, PyObject* args, PyObject* kwd
     self->nindex_ = 0;
     self->type_ = PyHoc::HocTopLevelInterpreter;
     self->iteritem_ = 0;
-    if (kwds && PyDict_Check(kwds)) {
-        PyObject* base = PyDict_GetItemString(kwds, "hocbase");
-        if (base) {
-            int ok = 0;
-            if (PyObject_TypeCheck(base, hocobject_type)) {
-                PyHocObject* hbase = (PyHocObject*) base;
-                if (hbase->type_ == PyHoc::HocFunction && hbase->sym_->type == TEMPLATE) {
-                    // printf("hocobj_new base %s\n", hbase->sym_->name);
-                    // remove the hocbase keyword since hocobj_call only allows
-                    // the "sec" keyword argument
-                    PyDict_DelItemString(kwds, "hocbase");
-                    PyObject* r = hocobj_call(hbase, args, kwds);
-                    if (!r) {
-                        Py_DECREF(subself);
-                        return NULL;
-                    }
-                    PyHocObject* rh = (PyHocObject*) r;
-                    self->type_ = rh->type_;
-                    self->ho_ = rh->ho_;
-                    hoc_obj_ref(self->ho_);
-                    Py_DECREF(r);
-                    ok = 1;
-                }
-            }
-            if (!ok) {
-                Py_DECREF(subself);
-                PyErr_SetString(PyExc_TypeError, "HOC base class not valid");
-                return NULL;
-            }
+
+    // if subtype is a subclass of some NEURON class, then one of its
+    // tp_mro's is in sym_to_type_map
+    for (Py_ssize_t i = 0; i < PyTuple_Size(subtype->tp_mro); i++) {
+        PyObject* item = PyTuple_GetItem(subtype->tp_mro, i);
+        auto symbol_result = type_to_sym_map.find((PyTypeObject*) item);
+        if (symbol_result != type_to_sym_map.end()) {
+            hbase = (PyHocObject*) hocobj_new(hocobject_type, 0, 0);
+            hbase->type_ = PyHoc::HocFunction;
+            hbase->sym_ = symbol_result->second;
+            break;
+        }
+    }
+
+    if (kwds && PyDict_Check(kwds) && (base = PyDict_GetItemString(kwds, "hocbase"))) {
+        if (PyObject_TypeCheck(base, hocobject_type)) {
+            hbase = (PyHocObject*) base;
+        }
+        else {
+            PyErr_SetString(PyExc_TypeError, "HOC base class not valid");
+            Py_DECREF(subself);
+            return NULL;
         }
+        PyDict_DelItemString(kwds, "hocbase");
     }
+
+    if (hbase and hbase->type_ == PyHoc::HocFunction && hbase->sym_->type == TEMPLATE) {
+        // printf("hocobj_new base %s\n", hbase->sym_->name);
+        // remove the hocbase keyword since hocobj_call only allows
+        // the "sec" keyword argument
+        PyObject* r = hocobj_call(hbase, args, kwds);
+        if (!r) {
+            Py_DECREF(subself);
+            return NULL;
+        }
+        PyHocObject* rh = (PyHocObject*) r;
+        self->type_ = rh->type_;
+        self->ho_ = rh->ho_;
+        hoc_obj_ref(self->ho_);
+        Py_DECREF(r);
```

Comment:
@ramcdougal is this correct? Should we decrement the ref of the object that was created?

Reply:
It seems like the hoc structure is transferred to `self`, so `r` may be discarded. It looks like this happened before as well. We could check if disabling that fixes the error?

Reply:
I don't see how it could be right... I had thought `hoc_obj_ref` incremented the Python ref count when the corresponding HocObject, but [it does not](https://github.com/neuronsimulator/nrn/blob/a5c36e9d746a52344d34eb9fa65da8a10e3dd95b/src/oc/hoc_oop.cpp#L2009).

@Helveg is right that this code was there before... but this was also a relatively unused feature.